EIPS/eip-4788.md

```diff
@@ -77,76 +79,165 @@ When verifying a block, execution clients **MUST** ensure the root value in the
 
 For a genesis block with no existing parent beacon block root the 32 zero bytes are used as a root placeholder.
 
-### EVM changes
+#### Beacon roots contract
 
-#### Block processing
+The beacon roots contract has two operations: `get` and `set`. The input itself is not used to determine which function to execute, for that the result of `caller` is used. If `caller` is equal to `SYSTEM_ADDRESS` then the operation to perform is `set`. Otherwise, `get`.
 
-At the start of processing any execution block where `block.timestamp >= FORK_TIMESTAMP` (i.e. before processing any transactions),
-write the parent beacon root provided in the block header into the storage of the contract at `HISTORY_STORAGE_ADDRESS`.
+##### `get`
 
-In order to bound the storage used by this precompile, two ring buffers are used: one to track the latest timestamp at a given index in the ring buffer and another to track
-the latest root at a given index.
+* If `caller` is equal to `SYSTEM_ADDRESS`, the contract must revert.
+* Callers provide the `timestamp` they are querying encoded as 32 bytes in big-endian format.
+* If the input is not exactly 32 bytes, the contract must revert.
+* Given `timestamp`, the contract computes the storage index in which the timestamp is stored by computing the modulo `timestamp % HISTORICAL_ROOTS_MODULUS` and reads the value.
+* If the `timestamp` does not match, the contract must revert.
+* Finally, the beacon root associated with the timestamp is accessed at returned to the user. It is stored at `timestamp % HISTORICAL_ROOTS_MODULUS + HISTORICAL_ROOTS_MODULUS`.
 
-To derive the index `timestamp_index` into the timestamp ring buffer, the timestamp (a 64-bit unsigned integer value) is reduced modulo `HISTORICAL_ROOTS_MODULUS`.
-To derive the index `root_index` into the root ring buffer, add `HISTORICAL_ROOTS_MODULUS` to the index into the timestamp ring buffer.
-Both resulting 64-bit unsigned integers should be encoded as 32 bytes in big-endian format when writing to the storage.
+##### `set`
 
-The timestamp from the header, encoded as 32 bytes in big-endian format, is the value to write behind the `timestamp_index`.
-The 32 bytes of the `parent_beacon_block_root` (as provided) are the value to write behind the `root_index`.
+* If `caller` is not equal to `SYSTEM_ADDRESS`, the contract must revert.
+* Caller provides the parent beacon block root as calldata to the contract.
+* Set the storage value at `header.timestamp % HISTORICAL_ROOTS_MODULUS` to be `header.timestamp`
```

Comment:
I kind of want to make the case for avoiding  a system-TX and just setting these storage values at the top of the block. Then the code is just `get` and we don't have to test/specify how to do system-TXs 

weaker held opinion than the deploy method

Reply:
I think if we're going to use a standard contract we should embrace it and allow it to have storing functionality.

Reply:
Wrote this already in discord, posting here too for visibility. 

With the spec being 'system call', it means a reference client can just do a simple call. A client implementor can still choose not to, and instead bypass the whole call and update the values directly. 

If the eip standardizes 'direct update', however, and omits the system-update path from the contract, then we remove all optionality, and force more special-case code into clients

Therefore I think the 'system call' approach is the best way forward.

Reply:
I think the EIP should mention that "direct update" is a valid/possible client implementation/optimization detail.

Reply:
having a system-tx would be a good generic functionality to have in long term imo